Objects/unicodeobject.c

```diff
@@ -9272,24 +9272,24 @@ tailmatch(PyObject *self,
     else
         offset = start;
 
-    if (PyUnicode_READ(kind_self, data_self, offset) ==
-        PyUnicode_READ(kind_sub, data_sub, 0) &&
-        PyUnicode_READ(kind_self, data_self, offset + end_sub) ==
-        PyUnicode_READ(kind_sub, data_sub, end_sub)) {
+    int match_last = PyUnicode_READ(kind_self, data_self, offset + end_sub) ==
+                        PyUnicode_READ(kind_sub, data_sub, end_sub);
+
+    if (match_last) {
+        if (end_sub==0) {
+            return 1;
+        }
         /* If both are of the same kind, memcmp is sufficient */
         if (kind_self == kind_sub) {
-            return ! memcmp((char *)data_self +
-                                (offset * PyUnicode_KIND(substring)),
-                            data_sub,
-                            PyUnicode_GET_LENGTH(substring) *
-                                PyUnicode_KIND(substring));
+            return ! memcmp((char *)data_self + (offset * kind_sub),
+                            data_sub, end_sub * kind_sub);
```

Comment:
For the last argument, the previous `memcmp` used `PyUnicode_GET_LENGTH(substring) * PyUnicode_KIND(substring)`, but this PR uses `end_sub * kind_sub` which computes to `(PyUnicode_GET_LENGTH(substring) - 1) * PyUnicode_KIND(substring)`. Perhaps it does not matter.

Reply:
We can compare from zero to `(PyUnicode_GET_LENGTH(substring) - 1)` because the last character has already been checked (this is the check `if (match_last)`. This makes the number of bytes to compare by `memcmp` a bit less. Note: the main performance gain in this PR is not due to this optimization, but due to avoiding the check with `PyUnicode_READ` on the first character.

Reply:
Got it; thanks!